### PR TITLE
[GOBBLIN-259] Support writing Kafka messages to db/table file path

### PR DESCRIPTION
**Pull request by zxcware**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title.
    - https://issues.apache.org/jira/browse/GOBBLIN-259

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
- Add a new write file path type `DB_TABLE`, which writes records from an `Extract` to folder '<namespace>/<table>'. 
- A gobblin job which uses the `KafkSource` and `AvroHdfsDataWriter` will write records to 'dbName/tableName' with the following configurations
```
extract.namespace=dbName
extract.table.name=tableName
writer.file.path.type=db_table
```

### Tests
- [x] My PR adds the following unit tests
- `WriterUtilsTest#testGetWriterFilePath`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by zxcware** [comment]
@yukuai518 @htran1 Please review. Thanks.

File: gobblin-utility/src/test/java/org/apache/gobblin/util/WriterUtilsTest.java [hunk 1]
```diff
@@ -93,13 +93,24 @@ public void testGetDataPublisherFinalOutputDir() {
...
     state.setProp(ConfigurationKeys.WRITER_FILE_PATH + ".0", TEST_WRITER_FILE_PATH);
     Assert.assertEquals(WriterUtils.getWriterFilePath(state, 1, 1), TEST_WRITER_FILE_PATH);
+
+    state.removeProp(ConfigurationKeys.WRITER_FILE_PATH);
+
+    state.setProp(ConfigurationKeys.WRITER_FILE_PATH_TYPE, "tablename");
+    Assert.assertEquals(WriterUtils.getWriterFilePath(state, 0, 0), new Path(table));
+
+    state.setProp(ConfigurationKeys.WRITER_FILE_PATH_TYPE, "db_table");
+    Assert.assertEquals(WriterUtils.getWriterFilePath(state, 0, 0), new Path(namespace + "/" + table));
```
Comment (htran1) [inline gobblin-utility/src/test/java/org/apache/gobblin/util/WriterUtilsTest.java]: Use `Path.SEPARATOR` here since you are using `Path.SEPARATOR` when constructing the name.

File: gobblin-api/src/main/java/org/apache/gobblin/source/workunit/WorkUnit.java [hunk 2]
```diff
@@ -82,6 +82,7 @@ public WorkUnit(SourceState state, Extract extract) {
 
     if (extract != null) {
       this.extract = extract;
+      super.addAll(extract);
```
Comment (htran1) [inline gobblin-api/src/main/java/org/apache/gobblin/source/workunit/WorkUnit.java]: Why this change instead of leaving the code that propagates only the table name to the workunit? This may bloat up the state which is already large for gobblin-kafka. Also you would have to add this the constructors that don't call this one. Unless there is a reason for this, I would prefer to leave it the way it was to avoid unexpected side effects.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/WriterUtils.java [hunk 3]
```diff
@@ -170,6 +175,19 @@ private static WriterFilePathType getWriterFilePathType(State state) {
...
   /**
+   * Creates {@link Path} for case {@link WriterFilePathType#DB_TABLE} with configurations
+   * {@link ConfigurationKeys#EXTRACT_NAMESPACE_NAME_KEY} and {@link ConfigurationKeys#EXTRACT_TABLE_NAME_KEY}
+   * @param state
+   * @return a path
+   */
+  public static Path getDbTableWriterFilePath(State state) {
+    Preconditions.checkArgument(state.contains(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY));
+    Preconditions.checkArgument(state.contains(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY));
+    return new Path(state.getProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY) + Path.SEPARATOR + state
```
Comment (yukuai518) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/WriterUtils.java]: What does EXTRACT_NAMESPACE_NAME_KEY mean in reality? Is it a DB name? or the DB name would be included in the topic.name? It's better to use specific name like EXTRACT_DB_NAME_KEY if possible because EXTRACT_NAMESPACE_NAME_KEY is something like "com.linkedin" or "gobblin.copy".

Comment (zxcware) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/WriterUtils.java]: EXTRACT_NAMESPACE_NAME_KEY defines a general domain of a table. It can be used as a DB name. A topic name is not necessary a fully qualified table name. A fully qualified namespace can have parents, like "com.linkedin.dbName". So we don't need to create a specific configuration key for DB, using the general one is good enough. 

To avoid confusing, I'll rename the file path type to be `NAMESPACE_TABLE`.
